File: leet-code/39. Combination Sum/CombinationSum39_2.java
package com.example;

import java.util.ArrayList;
import java.util.List;

public class CombinationSum39_2 {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        return getCombinationSum(candidates, target, new ArrayList<>(), new ArrayList<>(), 0);
    }

    private List<List<Integer>> getCombinationSum(int[] candidates,
                                                  int target,
                                                  List<List<Integer>> combinations,
                                                  List<Integer> comb,
                                                  int idx) {
        int sums = getSums(comb);

        if (sums >= target) {
            if (sums == target) {
                combinations.add(comb.stream().toList());
            }
            return combinations;
        }

        for (int i = idx; i < candidates.length; i++) {
            comb.add(candidates[i]);
            getCombinationSum(candidates, target, combinations, comb, i);
            comb.remove(comb.size() - 1);
        }
        return combinations;
    }

    private int getSums(List<Integer> combination) {
        return combination
                .stream()
                .reduce(0, Integer::sum);
    }
}

File: leet-code/39. Combination Sum/CombinationSum39_2.kt
package com.example

class CombinationSum39_2 {
    fun combinationSum(candidates: IntArray, target: Int): List<List<Int>> {
        var sums = 0

        fun recursion(
            combs: MutableList<List<Int>>,
            comb: MutableList<Int>,
            idx: Int,
        ): List<List<Int>> {
            if (sums >= target) {
                if (sums == target) {
                    combs.add(comb.toList())
                }
                return combs
            }
            for (i in idx until candidates.size) {
                comb.add(candidates[i])
                sums += candidates[i]
                recursion(combs, comb, i)
                sums -= candidates[i]
                comb.removeLast()
            }
            return combs
        }

        return recursion(mutableListOf(), mutableListOf(), 0)
    }
}
